packages/vite/src/node/plugin.ts

```diff
@@ -182,15 +184,21 @@ export interface Plugin<A = any> extends RollupPlugin<A> {
     (
       this: PluginContext,
       id: string,
-      options?: { ssr?: boolean; environment?: string },
+      options?: {
+        ssr?: boolean
+        environment?: ModuleExecutionEnvironment | BuildEnvironment
+      },
     ) => Promise<LoadResult> | LoadResult
   >
   transform?: ObjectHook<
     (
       this: TransformPluginContext,
       code: string,
       id: string,
-      options?: { ssr?: boolean; environment?: string },
+      options?: {
+        ssr?: boolean
+        environment?: ModuleExecutionEnvironment | BuildEnvironment
```

Comment:
Can we guarantee it exists somehow?

Reply:
The problem is that during scan, we didn't pass the moduleGraph. See here https://github.com/vitejs/vite/blob/d2839bc63e50cf211eea72081c0ac595d37088e0/packages/vite/src/node/optimizer/scan.ts#L212

So I didn't pass the environment either. I thought that it may make the scan slower if we need to update info in a module graph that we are going to throw away. But I think it is wrong that we don't pass an environment, more once we will have environment specific config. I'll check what needs to be done here. I think we only need the plugins for `resolveId` that shouldn't mess with the moduleGraph, so maybe we can create a new temporal browser environment with a new module graph and pass that. I'll check this out, it would be a lot better if there is always an environment defined.

Reply:
Let's see if there isn't some internalish use cases of these hooks in the ecosystem, but for now, core works with `options` and `environment` being required for hooks. See https://github.com/vitejs/vite/pull/16129/commits/e30b858035a840e94ad00a1597b08b51971c7f13

I fixed some issues were I missed passing the environment, so even if we revert, it was worth it. We currently have a design issue because an environment needs a `resolveId` function when it is created, that should be implemented as `container.resolveId(id, undefined, { environment })`. I'll rework this next so we can simplify the code.

Reply:
It seems we'll need to relax the types. See [failing test for Vitest](https://github.com/vitejs/vite-ecosystem-ci/actions/runs/8248284624/job/22558283515). We could leave them as is for a bit more if they help us work with this draft though.

Reply:
I already moved back to optional `options` and `environment`. If not, we were forced to have a new `GlobalEnvironment` that would be used when we don't have a server available (in the plugin containers created during config resolution for example).
Still, this was quite helpful to identify issues. And things looks better after https://github.com/vitejs/vite/pull/16129/commits/fcebb7dbaf46592bc4f4064888d4bcdf439f466c